File: scripts/rrt_star/rrt_star_2d_yaw.py
import numpy as np
import random
import matplotlib.pyplot as plt

class Node:
    def __init__(self, point):
        self.point = np.array(point)  # Including yaw as the third dimension
        self.parent = None
        self.cost = 0

def rrt_star(start, goal, obstacles, dim_x=10, dim_y=10, max_iter=1000, step_size=0.1, radius=2, tolerance=0.3):
    tree = [Node(start)]
    goal_reached_nodes = []  # Keep track of nodes that reach the goal
    path = None
    
    for _ in range(max_iter):
        rand_point = sample_space(dim_x, dim_y)
        nearest_node = find_nearest(tree, rand_point)
        new_node = steer(nearest_node, rand_point, step_size)
        
        if not collides(new_node.point[:2], obstacles):  # Pass only x and y components
            nearby_nodes = find_nearby(tree, new_node, radius)
            new_node = choose_parent(new_node, nearby_nodes)
            tree.append(new_node)
            rewire(tree, new_node, nearby_nodes, radius)
            
            if np.linalg.norm(new_node.point[:2] - goal[:2]) <= tolerance:  # Adjusted condition
                print("Goal reached!")

                # Backtrack with the node having the smallest cost among those that reached the goal
                goal_reached_nodes.append(new_node)
                min_cost_node = min(goal_reached_nodes, key=lambda node: node.cost)
                path, next_best_node = backtrack_path_node(min_cost_node)
                
                # Calculate yaw angles for each node based on the path
                for i in range(len(path)-1):
                    path[i][2] = calculate_yaw_angle(Node(path[i]), Node(path[i+1]))
                    if i == len(path)-1:
                        path[i+1][2] = goal[2:]
    
    return tree, path


def sample_space(dim_x,dim_y):
    rand_x = random.random() * dim_x #* 2 * dim_x - dim_x
    rand_y = random.random() * dim_y #* 2 * dim_y - dim_y
    return np.array([rand_x, rand_y])

def find_nearest(tree, point):
    distances = [np.linalg.norm(node.point[:2] - point[:2]) for node in tree]  # Consider only x and y components
    nearest_index = np.argmin(distances)
    return tree[nearest_index]

def steer(from_node, to_point, step_size):
    if np.linalg.norm(to_point - from_node.point[:2]) < step_size:
        new_point = to_point
    else:
        direction = (to_point - from_node.point[:2]) / np.linalg.norm(to_point - from_node.point[:2])
        new_point = from_node.point[:2] + step_size * direction
    return Node(np.concatenate([new_point, from_node.point[2:]]))  # Preserve yaw angle

def collides(point, obstacles):
    for obstacle in obstacles:
        if np.linalg.norm(point[:2] - obstacle['point']) < obstacle['radius'] + 0.5:
            return True
    return False

def find_nearby(tree, point, radius):
    tree_points = np.array([node.point for node in tree])
    distances = np.linalg.norm(tree_points[:, :2] - point.point[:2], axis=1)  # Consider only x and y components
    nearby_nodes = [node for node, dist in zip(tree, distances) if dist < radius]
    return nearby_nodes

def choose_parent(point, nearby_nodes):
    min_cost = float('inf')
    parent = None
    for node in nearby_nodes:
        # Include yaw angle difference in the cost computation
        cost = node.cost + np.linalg.norm(node.point[:2] - point.point[:2])
        if cost < min_cost:
            min_cost = cost
            parent = node
    point.parent = parent
    point.cost = min_cost
    return point

def rewire(tree, new_node, nearby_nodes, radius):
    for node in nearby_nodes:
        if node.cost > new_node.cost + np.linalg.norm(node.point[:2] - new_node.point[:2]):
            node.parent = new_node
            node.cost = new_node.cost + np.linalg.norm(node.point[:2] - new_node.point[:2])

def backtrack_path_node(node):
    path = []
    next_best_node = node
    while node:
        path.append(node.point)
        node = node.parent
        if node:
            if node.parent: # Exclude root node
                if node.cost < next_best_node.cost:
                    next_best_node = node
    path.reverse()
    return path, next_best_node

def calculate_yaw_angle(node1, node2):
    dx = node2.point[0] - node1.point[0]
    dy = node2.point[1] - node1.point[1]
    return np.arctan2(dy, dx)

def plot_tree(tree, start, goal, obstacles, path=None):
    plt.figure()
    # Plot tree edges
    for node in tree:
        if node.parent:
            plt.plot([node.point[0], node.parent.point[0]], [node.point[1], node.parent.point[1]], 'k-', linewidth=0.5)
    # Plot obstacles
    for obstacle in obstacles:
        circle = plt.Circle(obstacle['point'][:2], obstacle['radius'], color='r')
        plt.gca().add_patch(circle)
    # Plot start and goal points
    plt.scatter(start[0], start[1], color='g', marker='o', label='Start')
    plt.scatter(goal[0], goal[1], color='b', marker='o', label='Goal')
    # Plot final path (if provided)
    if path:
        plt.plot([p[0] for p in path], [p[1] for p in path], 'b-', linewidth=2, label='Path')
    # Labels and formatting
    plt.xlabel('X')
    plt.ylabel('Y')
    plt.title('RRT*')
    plt.legend()
    plt.axis('equal')
    plt.show()

'''
# Example usage
start = np.array([1, 1, 0])
goal = np.array([9, 9, 0])
dim_x = 10
dim_y = 10
obstacles = [{'point': np.array([5, 5]), 'radius': 1}]
tree, path = rrt_star(start, goal, obstacles, dim_x=dim_x, dim_y=dim_y, step_size=0.5, radius=2)
#print(path)
plot_tree(tree, start, goal, obstacles, path)
'''

File: src/RRT/rrt_star_2d_yaw.cpp
#include "motion_planning_python/RRT/rrt_star_2d_yaw.h"
#include <iostream>
#include <random>
#include <algorithm>
#include <limits>

namespace rrt_star_2d {

Node::Node(const Eigen::Vector3d& p) : point(p), parent(nullptr), cost(0), gain(0), score(0) {}

Eigen::Vector2d sampleSpace(double dim_x, double dim_y) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::uniform_real_distribution<double> dis_x(0.0, dim_x);
    std::uniform_real_distribution<double> dis_y(0.0, dim_y);
    double rand_x = dis_x(gen);
    double rand_y = dis_y(gen);
    return Eigen::Vector2d(rand_x, rand_y);
}

Node* findNearest(std::vector<Node*>& tree, const Eigen::Vector2d& point) {
    double minDist = std::numeric_limits<double>::max();
    Node* nearestNode = nullptr;
    for (Node* node : tree) {
        double distance = (node->point.head<2>() - point).norm();
        if (distance < minDist) {
            minDist = distance;
            nearestNode = node;
        }
    }
    return nearestNode;
}

Node* steer(Node* fromNode, const Eigen::Vector2d& toPoint, double stepSize) {
    double dist = (toPoint - fromNode->point.head<2>()).norm();
    if (dist < stepSize) {
        return new Node(Eigen::Vector3d(toPoint.x(), toPoint.y(), fromNode->point.z()));
    } else {
        Eigen::Vector2d direction = (toPoint - fromNode->point.head<2>()).normalized();
        Eigen::Vector2d newPoint = fromNode->point.head<2>() + stepSize * direction;
        return new Node(Eigen::Vector3d(newPoint.x(), newPoint.y(), fromNode->point.z()));
    }
}

bool collides(const std::vector<double>& point, const std::vector<std::pair<Eigen::Vector2d, double>>& obstacles) {
    for (const auto& obstacle : obstacles) {
        double distance = (point[0] - obstacle.first[0]) * (point[0] - obstacle.first[0]) +
                          (point[1] - obstacle.first[1]) * (point[1] - obstacle.first[1]);
        if (distance < obstacle.second * obstacle.second + 0.5) {
            return true;
        }
    }
    return false;
}

std::vector<Node*> findNearby(std::vector<Node*>& tree, Node* point, double radius) {
    std::vector<Node*> nearbyNodes;
    for (Node* node : tree) {
        double distance = (node->point.head<2>() - point->point.head<2>()).norm();
        if (distance < radius) {
            nearbyNodes.push_back(node);
        }
    }
    return nearbyNodes;
}

Node* chooseParent(Node* point, const std::vector<Node*>& nearbyNodes) {
    double minCost = std::numeric_limits<double>::infinity();
    Node* parent = nullptr;
    for (Node* node : nearbyNodes) {
        double cost = node->cost + (node->point.head<2>() - point->point.head<2>()).norm();
        if (cost < minCost) {
            minCost = cost;
            parent = node;
        }
    }
    point->parent = parent;
    point->cost = minCost;
    return point;
}

void rewire(std::vector<Node*>& tree, Node* new_node, const std::vector<Node*>& nearby_nodes, double radius) {
    for (const auto& node : nearby_nodes) {
        double new_cost = new_node->cost + (node->point.head<2>() - new_node->point.head<2>()).norm();
        if (new_cost < node->cost) {
            node->parent = new_node;
            node->cost = new_cost;
        }
    }
}

double calculateYawAngle(const Node* node1, const Node* node2) {
    double dx = node2->point.x() - node1->point.x();
    double dy = node2->point.y() - node1->point.y();
    return std::atan2(dy, dx);
}

std::pair<std::vector<Eigen::Vector3d>, Node*> backtrackPathNode(Node* node) {
    std::vector<Eigen::Vector3d> path;
    Node* nextBestNode = node;
    while (node) {
        path.push_back(Eigen::Vector3d(node->point));
        node = node->parent;
        if (node && node->parent && node->cost < nextBestNode->cost) {
            nextBestNode = node;
        }
    }
    std::reverse(path.begin(), path.end());
    return {path, nextBestNode};
}

std::pair<std::vector<Node*>, std::vector<Eigen::Vector3d>> rrtStar(const Eigen::Vector3d& start, const Eigen::Vector3d& goal,
                                                                    const std::vector<std::pair<Eigen::Vector2d, double>>& obstacles,
                                                                    double dim_x, double dim_y, int max_iter,
                                                                    double step_size, double radius, double tolerance) {
    std::vector<Node*> tree = {new Node(start)};
    std::vector<Node*> goalReachedNodes;
    std::vector<Eigen::Vector3d> path;
    Node* minCostNode = nullptr; // Declaration of minCostNode variable

    for (int i = 0; i < max_iter; ++i) {
        Eigen::Vector2d randPoint = sampleSpace(dim_x, dim_y);
        Node* nearestNode = findNearest(tree, randPoint);
        Node* newNode = steer(nearestNode, randPoint, step_size);
        
        if (!collides({newNode->point.x(), newNode->point.y()}, obstacles)) {
            std::vector<Node*> nearbyNodes = findNearby(tree, newNode, radius);
            newNode = chooseParent(newNode, nearbyNodes);
            tree.push_back(newNode);
            rewire(tree, newNode, nearbyNodes, radius);
            
            if ((newNode->point.head<2>() - goal.head<2>()).norm() <= tolerance) {
                std::cout << "Goal reached!" << std::endl;

                goalReachedNodes.push_back(newNode);
                minCostNode = *std::min_element(goalReachedNodes.begin(), goalReachedNodes.end(),
                                                      [](Node* node1, Node* node2) { return node1->cost < node2->cost; });
                auto [p, nextBestNode] = backtrackPathNode(minCostNode);
                path = p;

                for (size_t i = 0; i < path.size() - 1; ++i) {
                    path[i][2] = calculateYawAngle(new Node(path[i]), new Node(path[i + 1]));
                    if (i == path.size() - 2) {
                        path[i + 1][2] = goal[2];
                    }
                }
                return {tree, path};
            }
        }
    }
    return {tree, path};
}


/*int main() {
    Eigen::Vector3d start(1, 1, 0);
    Eigen::Vector3d goal(9, 9, 0);
    double dim_x = 10;
    double dim_y = 10;
    std::vector<std::pair<Eigen::Vector2d, double>> obstacles = { {Eigen::Vector2d(5, 5), 1} };
    auto [tree, path] = rrtStar(start, goal, obstacles, dim_x, dim_y, 1000, 0.5, 2, 0.5);
    return 0;
}*/

}  // namespace rrt_star
